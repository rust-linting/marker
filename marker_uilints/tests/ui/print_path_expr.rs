pub trait A {
    fn a() {
        println!("a()");
    }
}

pub trait B {
    type CoolTy: A;

    fn assoc_item_ty(cool: Self::CoolTy) -> Self::CoolTy {
        cool
    }
}

struct Magic;
struct Sparkles;

impl A for Sparkles {}

impl B for Magic {
    type CoolTy = Sparkles;
}

fn rand(_seed: u32) -> u16 {
    5
}

pub fn main() {
    let _print_path = rand(7);
    let _print_path = Vec::<u32>::new();
    let _print_path = <Magic as B>::CoolTy::a();
    let _print_path = <<Magic as B>::CoolTy as A>::a();

    let var = 1;
    let _print_path = var;
}
