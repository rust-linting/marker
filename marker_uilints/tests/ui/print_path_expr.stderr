warning: print test
  --> $DIR/print_path_expr.rs:29:5
   |
29 |     let _print_path = rand(7);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Call(
               CallExpr {
                   data: CommonExprData {
                       _lifetime: PhantomData<&()>,
                       id: ExprId(..),
                       span: SpanId(..),
                   },
                   func: Path(
                       PathExpr {
                           data: CommonExprData {
                               _lifetime: PhantomData<&()>,
                               id: ExprId(..),
                               span: SpanId(..),
                           },
                           path: AstQPath {
                               self_ty: None,
                               path_ty: None,
                               path: AstPath {
                                   segments: [
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "rand",
                                               span: $DIR/print_path_expr.rs:29:23 - 29:27,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                   ],
                               },
                               target: Item(
                                   ItemId(..),
                               ),
                           },
                       },
                   ),
                   args: [
                       IntLit(
                           IntLitExpr {
                               data: CommonExprData {
                                   _lifetime: PhantomData<&()>,
                                   id: ExprId(..),
                                   span: SpanId(..),
                               },
                               value: 7,
                               suffix: None,
                           },
                       ),
                   ],
               },
           )
   = note: `#[warn(marker::marker_uilints::test_lint)]` on by default

warning: print test
  --> $DIR/print_path_expr.rs:30:5
   |
30 |     let _print_path = Vec::<u32>::new();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Call(
               CallExpr {
                   data: CommonExprData {
                       _lifetime: PhantomData<&()>,
                       id: ExprId(..),
                       span: SpanId(..),
                   },
                   func: Path(
                       PathExpr {
                           data: CommonExprData {
                               _lifetime: PhantomData<&()>,
                               id: ExprId(..),
                               span: SpanId(..),
                           },
                           path: AstQPath {
                               self_ty: None,
                               path_ty: Some(
                                   Path(
                                       PathTy {
                                           data: CommonSynTyData {
                                               _lifetime: PhantomData<&()>,
                                               span: SpanId(..),
                                           },
                                           path: AstQPath {
                                               self_ty: None,
                                               path_ty: None,
                                               path: AstPath {
                                                   segments: [
                                                       AstPathSegment {
                                                           ident: Ident {
                                                               name: "Vec",
                                                               span: $DIR/print_path_expr.rs:30:23 - 30:26,
                                                           },
                                                           generics: GenericArgs {
                                                               args: [
                                                                   Ty(
                                                                       TyArg {
                                                                           ty: Num(
                                                                               NumTy {
                                                                                   data: CommonSynTyData {
                                                                                       _lifetime: PhantomData<&()>,
                                                                                       span: SpanId(..),
                                                                                   },
                                                                                   numeric_kind: U32,
                                                                               },
                                                                           ),
                                                                       },
                                                                   ),
                                                               ],
                                                           },
                                                       },
                                                   ],
                                               },
                                               target: Item(
                                                   ItemId(..),
                                               ),
                                           },
                                       },
                                   ),
                               ),
                               path: AstPath {
                                   segments: [
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "Vec",
                                               span: $DIR/print_path_expr.rs:30:23 - 30:26,
                                           },
                                           generics: GenericArgs {
                                               args: [
                                                   Ty(
                                                       TyArg {
                                                           ty: Num(
                                                               NumTy {
                                                                   data: CommonSynTyData {
                                                                       _lifetime: PhantomData<&()>,
                                                                       span: SpanId(..),
                                                                   },
                                                                   numeric_kind: U32,
                                                               },
                                                           ),
                                                       },
                                                   ),
                                               ],
                                           },
                                       },
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "new",
                                               span: $DIR/print_path_expr.rs:30:35 - 30:38,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                   ],
                               },
                               target: Item(
                                   ItemId(..),
                               ),
                           },
                       },
                   ),
                   args: [],
               },
           )

warning: print test
  --> $DIR/print_path_expr.rs:31:5
   |
31 |     let _print_path = <Magic as B>::CoolTy::a();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Call(
               CallExpr {
                   data: CommonExprData {
                       _lifetime: PhantomData<&()>,
                       id: ExprId(..),
                       span: SpanId(..),
                   },
                   func: Path(
                       PathExpr {
                           data: CommonExprData {
                               _lifetime: PhantomData<&()>,
                               id: ExprId(..),
                               span: SpanId(..),
                           },
                           path: AstQPath {
                               self_ty: None,
                               path_ty: Some(
                                   Path(
                                       PathTy {
                                           data: CommonSynTyData {
                                               _lifetime: PhantomData<&()>,
                                               span: SpanId(..),
                                           },
                                           path: AstQPath {
                                               self_ty: Some(
                                                   Path(
                                                       PathTy {
                                                           data: CommonSynTyData {
                                                               _lifetime: PhantomData<&()>,
                                                               span: SpanId(..),
                                                           },
                                                           path: AstQPath {
                                                               self_ty: None,
                                                               path_ty: None,
                                                               path: AstPath {
                                                                   segments: [
                                                                       AstPathSegment {
                                                                           ident: Ident {
                                                                               name: "Magic",
                                                                               span: $DIR/print_path_expr.rs:31:24 - 31:29,
                                                                           },
                                                                           generics: GenericArgs {
                                                                               args: [],
                                                                           },
                                                                       },
                                                                   ],
                                                               },
                                                               target: Item(
                                                                   ItemId(..),
                                                               ),
                                                           },
                                                       },
                                                   ),
                                               ),
                                               path_ty: None,
                                               path: AstPath {
                                                   segments: [
                                                       AstPathSegment {
                                                           ident: Ident {
                                                               name: "B",
                                                               span: $DIR/print_path_expr.rs:31:33 - 31:34,
                                                           },
                                                           generics: GenericArgs {
                                                               args: [],
                                                           },
                                                       },
                                                       AstPathSegment {
                                                           ident: Ident {
                                                               name: "CoolTy",
                                                               span: $DIR/print_path_expr.rs:31:37 - 31:43,
                                                           },
                                                           generics: GenericArgs {
                                                               args: [],
                                                           },
                                                       },
                                                   ],
                                               },
                                               target: Item(
                                                   ItemId(..),
                                               ),
                                           },
                                       },
                                   ),
                               ),
                               path: AstPath {
                                   segments: [
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "B",
                                               span: $DIR/print_path_expr.rs:31:33 - 31:34,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "CoolTy",
                                               span: $DIR/print_path_expr.rs:31:37 - 31:43,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "a",
                                               span: $DIR/print_path_expr.rs:31:45 - 31:46,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                   ],
                               },
                               target: Item(
                                   ItemId(..),
                               ),
                           },
                       },
                   ),
                   args: [],
               },
           )

warning: print test
  --> $DIR/print_path_expr.rs:32:5
   |
32 |     let _print_path = <<Magic as B>::CoolTy as A>::a();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Call(
               CallExpr {
                   data: CommonExprData {
                       _lifetime: PhantomData<&()>,
                       id: ExprId(..),
                       span: SpanId(..),
                   },
                   func: Path(
                       PathExpr {
                           data: CommonExprData {
                               _lifetime: PhantomData<&()>,
                               id: ExprId(..),
                               span: SpanId(..),
                           },
                           path: AstQPath {
                               self_ty: Some(
                                   Path(
                                       PathTy {
                                           data: CommonSynTyData {
                                               _lifetime: PhantomData<&()>,
                                               span: SpanId(..),
                                           },
                                           path: AstQPath {
                                               self_ty: Some(
                                                   Path(
                                                       PathTy {
                                                           data: CommonSynTyData {
                                                               _lifetime: PhantomData<&()>,
                                                               span: SpanId(..),
                                                           },
                                                           path: AstQPath {
                                                               self_ty: None,
                                                               path_ty: None,
                                                               path: AstPath {
                                                                   segments: [
                                                                       AstPathSegment {
                                                                           ident: Ident {
                                                                               name: "Magic",
                                                                               span: $DIR/print_path_expr.rs:32:25 - 32:30,
                                                                           },
                                                                           generics: GenericArgs {
                                                                               args: [],
                                                                           },
                                                                       },
                                                                   ],
                                                               },
                                                               target: Item(
                                                                   ItemId(..),
                                                               ),
                                                           },
                                                       },
                                                   ),
                                               ),
                                               path_ty: None,
                                               path: AstPath {
                                                   segments: [
                                                       AstPathSegment {
                                                           ident: Ident {
                                                               name: "B",
                                                               span: $DIR/print_path_expr.rs:32:34 - 32:35,
                                                           },
                                                           generics: GenericArgs {
                                                               args: [],
                                                           },
                                                       },
                                                       AstPathSegment {
                                                           ident: Ident {
                                                               name: "CoolTy",
                                                               span: $DIR/print_path_expr.rs:32:38 - 32:44,
                                                           },
                                                           generics: GenericArgs {
                                                               args: [],
                                                           },
                                                       },
                                                   ],
                                               },
                                               target: Item(
                                                   ItemId(..),
                                               ),
                                           },
                                       },
                                   ),
                               ),
                               path_ty: None,
                               path: AstPath {
                                   segments: [
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "A",
                                               span: $DIR/print_path_expr.rs:32:48 - 32:49,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                       AstPathSegment {
                                           ident: Ident {
                                               name: "a",
                                               span: $DIR/print_path_expr.rs:32:52 - 32:53,
                                           },
                                           generics: GenericArgs {
                                               args: [],
                                           },
                                       },
                                   ],
                               },
                               target: Item(
                                   ItemId(..),
                               ),
                           },
                       },
                   ),
                   args: [],
               },
           )

warning: print test
  --> $DIR/print_path_expr.rs:35:5
   |
35 |     let _print_path = var;
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Path(
               PathExpr {
                   data: CommonExprData {
                       _lifetime: PhantomData<&()>,
                       id: ExprId(..),
                       span: SpanId(..),
                   },
                   path: AstQPath {
                       self_ty: None,
                       path_ty: None,
                       path: AstPath {
                           segments: [
                               AstPathSegment {
                                   ident: Ident {
                                       name: "var",
                                       span: $DIR/print_path_expr.rs:35:23 - 35:26,
                                   },
                                   generics: GenericArgs {
                                       args: [],
                                   },
                               },
                           ],
                       },
                       target: Var(
                           VarId(..),
                       ),
                   },
               },
           )

warning: 5 warnings emitted

